albatross/server.py

```diff
@@ -64,41 +108,33 @@ def _parse_header_lines(self, lines):
         return ImmutableCaselessDict(*headers.items())
 
     async def _parse_request(self, request_reader):
-        request_line = await request_reader.readline()
-        request_line = request_line.decode()
-        method, url_string, _ = request_line.split(' ', 2)
-        method = method.upper()
-        url = parse.urlparse(url_string)
+        # Read next event and see if it is a request
 
-        header_lines = []
-        while True:
-            l = await request_reader.readline()
-            if l == b'\r\n':
-                break
-            header_lines.append(l.decode())
+        event = await self.next_event(request_reader)
 
-        headers = self._parse_header_lines(header_lines)
+        if type(event) is h11.Request:
 
-        raw_body = None
-        if method in {'POST', 'PUT'}:
-            body_parts = []
-            content_length = int(headers['Content-Length'])
-            while content_length > 0:
-                chunk_size = min(content_length, self.max_read_chunk)
-                body = await request_reader.read(chunk_size)
-                content_length -= len(body)
-                body_parts.append(body)
-            raw_body = b''.join(body_parts)
-            del body_parts
+            # Gather request parts
+            path = event.target.decode("ascii")
+            method = event.method.decode("ascii")
+            headers = [(name.decode("ascii"), value.decode("ascii"))
+                       for (name, value) in event.headers]
+            query = ""
+            raw_body = ""
 
-        path = url.path
-        query = url.query
+            # Now get any data
+            while True:
+                event = await self.next_event(request_reader)
+                if type(event) is h11.EndOfMessage:
+                    break
+                assert type(event) is h11.Data
+                raw_body += event.data.decode("ascii")
```

Comment:
its bad to construct strings like this. You want either io.StringIO or an array and then a ''.join([...]